author	Par-Gunnar Hjalmdahl <par-gunnar.hjalmdahl@stericsson.com>
commit	2b44cd2fba6e9a8590f30e68db0f6b92e8fcdb94
	Wed, 15 Aug 2012 12:42:06 +0000 (15:42 +0300)

network: Check full BNEP UUID

This patch fixes an issue where only the 2 bytes containing the service
ID was checked from the BNEP UUID. Fixes behavior for BT test cases
TP/PAN/MISC/UUID/BV-01-C & TP/PAN/MISC/UUID/BV-02-C.

diff -cr bluez-4.99.pan.orig/network/server.c bluez-4.99.pan/network/server.c
*** bluez-4.99.pan.orig/network/server.c	Mon Mar  5 04:03:57 2012
--- bluez-4.99.pan/network/server.c	Fri Oct 12 06:47:10 2012
***************
*** 315,321 ****
--- 315,324 ----
  static uint16_t bnep_setup_decode(struct bnep_setup_conn_req *req,
  				uint16_t *dst_role, uint16_t *src_role)
  {
+ 	const uint8_t bt_base[] = { 0x00, 0x00, 0x10, 0x00, 0x80, 0x00,
+ 				0x00, 0x80, 0x5F, 0x9B, 0x34, 0xFB };
  	uint8_t *dest, *source;
+ 	uint32_t val;
  
  	dest = req->service;
  	source = req->service + req->uuid_size;
***************
*** 325,334 ****
  		*dst_role = bt_get_be16(dest);
  		*src_role = bt_get_be16(source);
  		break;
- 	case 4: /* UUID32 */
  	case 16: /* UUID128 */
! 		*dst_role = bt_get_be32(dest);
! 		*src_role = bt_get_be32(source);
  		break;
  	default:
  		return BNEP_CONN_INVALID_SVC;
--- 328,354 ----
  		*dst_role = bt_get_be16(dest);
  		*src_role = bt_get_be16(source);
  		break;
  	case 16: /* UUID128 */
! 		/* Check that the bytes in the UUID, except the service ID
! 		 * itself, are correct. The service ID is checked in
! 		 * bnep_setup_chk(). */
! 		if (memcmp(&dest[4], bt_base, sizeof(bt_base)) != 0)
! 			return BNEP_CONN_INVALID_DST;
! 		if (memcmp(&source[4], bt_base, sizeof(bt_base)) != 0)
! 			return BNEP_CONN_INVALID_SRC;
! 
! 		/* Intentional no-break */
! 
! 	case 4: /* UUID32 */
! 		val = bt_get_be32(dest);
! 		if (val > 0xffff)
! 			return BNEP_CONN_INVALID_DST;
! 		*dst_role = val;
! 
! 		val = bt_get_be32(source);
! 		if (val > 0xffff)
! 			return BNEP_CONN_INVALID_SRC;
! 		*src_role = val;
  		break;
  	default:
  		return BNEP_CONN_INVALID_SVC;
